db/db_impl.cc

```diff
@@ -869,15 +859,28 @@ ColumnFamilyHandle* DBImpl::DefaultColumnFamily() const {
   return default_cf_handle_;
 }
 
+// Async version of the entry point
+Status DBImpl::Get(const GetCallback& cb, const ReadOptions& options,
+                  ColumnFamilyHandle* column_family, const Slice& key,
+                  std::string* value) {
+  return async::DBImplGetContext::RequestGet(cb, this, options, column_family,
+      key, nullptr, value);
+}
+
 Status DBImpl::Get(const ReadOptions& read_options,
                    ColumnFamilyHandle* column_family, const Slice& key,
                    PinnableSlice* value) {
-  return GetImpl(read_options, column_family, key, value);
+  return async::DBImplGetContext::Get(this, read_options, column_family,
```

Comment:
Suggestion for the first step: Instead of jumping to async:: code even for the sync code-flow, wondering if the code would be much cleaner if we call the methods in async namespace only if the callback is non-null, i.e. async.
Obviously, we could try refactoring the code in DBImpl::GetImpl into small functions that still stay in DBImpl.

Reply:
@sagar0 The goal was to have the same code for sync and async. Naturally, then the single code would be in the same class in SOME namespace but the class can not be in more than one namespace. It is alright to have it in a different namespace but I've striven to avoid code duplication.